stdlib/LinearAlgebra/src/diagonal.jl

```diff
@@ -368,6 +368,15 @@ end
     return out
 end
 
+function (*)(Da::Diagonal, A::AbstractMatrix, Db::Diagonal)
+    _muldiag_size_check(Da, A)
+    _muldiag_size_check(A, Db)
+    dda = Da.diag
+    ddb = Db.diag
+    n = length(dda)
+    return broadcast(*, dda, A, reshape(ddb, 1, n))
```

Comment:
This is very minor, but calling `permutedims` will future-proof this for offset axes etc, and might be more readable:
```suggestion
    return broadcast(*, Da.diag, A, permutedims(Db.diag))
```

Reply:
Also, is it worth avoiding the `reshape` for `Db::Diagonal{<:Number}`? Doesn't matter for large arrays, but for small ones, 20% at size 10, or:
```
julia> A = randn(20, 20);
julia> D = Diagonal(randn(20));
julia> D2 = Diagonal(randn(20));

julia> @btime $D * $A * $D2;
  min 453.893 ns, mean 1.774 μs (2 allocations, 6.50 KiB)

julia> @btime star1($D, $A, $D2);  # as in PR, with reshape
  min 315.275 ns, mean 943.622 ns (3 allocations, 3.34 KiB)

julia> function star2(Da::Diagonal, A::AbstractMatrix, Db::Diagonal)
            LinearAlgebra._muldiag_size_check(Da, A)
            LinearAlgebra._muldiag_size_check(A, Db)
            return broadcast(*, Da.diag, A, transpose(Db.diag))
       end

julia> @btime star2($D, $A, $D2);  # the mean is noisy, but min is not
  min 281.432 ns, mean 898.237 ns (1 allocation, 3.25 KiB)
```

Reply:
What is the drawback of using `transpose` here? IMO that difference between using transpose v.s. reshape is quite small, but we can switch to `transpose` if we get that 20% on small array for free.

Reply:
It won't be right for a matrix of matrices:
```
julia> D = Diagonal([rand(2,2) for i in 1:3]);

julia> M = fill(@SMatrix(rand(2,2)), 3, 3);

julia> D * M * D ≈ star1(D, M, D)
true

julia> D * M * D ≈ star2(D, M, D)
false
```
So either you need two methods, or I guess you only apply the optimisation to arrays of numbers.

Reply:
I never know we have Diagonal with matrix of matrices :sweat_smile:. If this optimisation also works for matrix of matrices, maybe it's better to keep the `reshape` one (iirc people has different thoughts on what's a Number)

Reply:
I do think transpose ought to be safe, we have `transpose(x::Number) = x` which (IMO) sort-of specifies this? But whether the speedup is worth it I don't know. 

I get about half the speedup from this, but maybe less generic...
```
      function star3(Da::Diagonal, A::AbstractMatrix, Db::Diagonal)
            LinearAlgebra._muldiag_size_check(Da, A)
            LinearAlgebra._muldiag_size_check(A, Db)
            return broadcast(*, Da.diag, A, Base.ReshapedArray(Db.diag, (1,length(Db.diag)), ()))
       end
```